tests/test.rs

```diff
@@ -151,6 +153,219 @@ async fn test_transaction() {
+            *should_retry_query_count.lock().unwrap() += 1;
+            true
+        },
+    )
+    .await
+    .expect("Transaction should have succeeded");
+    assert_eq!(conn.transaction_depth().await.unwrap(), 0);
+
+    // The transaction closure should have been attempted twice, but
+    // we should have only asked whether or not to retry once -- after
+    // the first failure, but not the second.
+    assert_eq!(
+        *transaction_attempted_count.lock().unwrap(),
+        EXPECTED_ERR_COUNT + 1
+    );
+    assert_eq!(
+        *should_retry_query_count.lock().unwrap(),
+        EXPECTED_ERR_COUNT
+    );
+
+    test_end(crdb).await;
+}
+
+#[tokio::test]
+async fn test_transaction_automatic_retry_does_not_retry_non_retryable_errors() {
+    let crdb = test_start().await;
+
+    let manager = async_bb8_diesel::ConnectionManager::<PgConnection>::new(&crdb.pg_config().url);
+    let pool = bb8::Pool::builder().build(manager).await.unwrap();
+    let conn = pool.get().await.unwrap();
+
+    // Test a transaction that:
+    //
+    // Fails with a non-retryable error. It should exit immediately.
+    assert_eq!(conn.transaction_depth().await.unwrap(), 0);
+    assert_eq!(
+        conn.transaction_async_with_retry(
+            |_| async { Err::<(), _>(diesel::result::Error::NotFound) },
+            || async { panic!("Should not attempt to retry this operation") }
+        )
+        .await
+        .expect_err("Transaction should have failed"),
+        diesel::result::Error::NotFound,
+    );
+    assert_eq!(conn.transaction_depth().await.unwrap(), 0);
+
+    test_end(crdb).await;
+}
+
+#[tokio::test]
+async fn test_transaction_automatic_retry_nested_transactions_fail() {
+    let crdb = test_start().await;
+
+    let manager = async_bb8_diesel::ConnectionManager::<PgConnection>::new(&crdb.pg_config().url);
+    let pool = bb8::Pool::builder().build(manager).await.unwrap();
+    let conn = pool.get().await.unwrap();
+
+    // This outer transaction should succeed immediately...
+    assert_eq!(conn.transaction_depth().await.unwrap(), 0);
+    conn.transaction_async_with_retry(
+        |conn| async move {
+            // ... but this inner transaction should fail! We do not support
+            // retryable nested transactions.
+            let err = conn
+                .transaction_async_with_retry(
+                    |_| async {
+                        panic!("Shouldn't run");
+                        #[allow(unreachable_code)]
+                        Ok(())
+                    },
+                    || async { panic!("Shouldn't retry inner transaction") },
+                )
+                .await
+                .expect_err("Nested transaction should have failed");
+            assert_eq!(err, diesel::result::Error::AlreadyInTransaction);
+            Ok(())
+        },
+        || async { panic!("Shouldn't retry outer transaction") },
+    )
+    .await
+    .expect("Transaction should have succeeded");
```

Comment:
It took me a minute to understand why this returns `Ok`. IIUC, it's because we're catching the `AlreadyInTransaction` error in the nested transaction closure (and panicking otherwise), and returning `Ok(())` from _that_. It might be more obvious as a test, and as a model for how to use this code, if we assert that the outer closure returns an `Err` instead. It _looks_ like the `AlreadyInTransaction` error should be bubbled up in that case, which I think it pretty helpful to see if one were tempted to write nested transactions using this.

I _think_ the functionality verified by this test is equivalent in either case, so this is mostly a didactic point.

Reply:
I actually intended for the outer transaction to succeed, so we could be **super** explicit about it not returning an error for some unrelated reason. My concern with returning `AlreadyInTransaction` is that it's not immediately obvious to me "which layer of transaction" is returning that error.

Maybe I could return a type that's not just `Ok(())` to make this more obvious? Like an explicit struct named `OnlyReturnFromOuterTransaction`?

(I'm trying that in the latest commit, lemme know what you think)

Reply:
I think that idea sounds good. I was just wondering what the outermost caller sees if there _is_ a nested transaction, and making sure that it's obvious to them how to fix it. I think that is the case though, so I'm happy with however this shakes out.

Reply:
They see whatever the "inner" transaction sees -- if it chooses to propagate errors, they'll see the error from the inner transaction. If it swallows the error (like we do in the test), they'll see whatever they want to return.

I actually saw a *variant* of this when integrating https://github.com/oxidecomputer/omicron/pull/4487 . I didn't have a "transaction_async_with_retry" nested within itself, but I **did** have a `transaction_async` nested inside a `transaction_async_with_retry`.

In this case, CRDB returned an error to me at run-time.

(I can't add this error checking to `transaction_async`, because that would ban **all** nested transactions, which would break Omicron -- see https://github.com/oxidecomputer/customer-support/issues/46#issuecomment-1813505462 for more details)